pydantic_ai_slim/pydantic_ai/agent.py

```diff
@@ -219,8 +232,10 @@ async def run(
                 with _logfire.span('preparing model and tools {run_step=}', run_step=run_step):
                     agent_model = await self._prepare_model(model_used, deps)
 
+                model_settings = {**(self.model_settings or {}), **(model_settings or {})}
```

Comment:
Since these can't change during iteration, move them out of the loop. Also probably worth have a method for this like

```py
def _build_model_settings(self, model_settings: ModelSettings | None) -> ModelSettings | None:
    if self.model_settings and model_settings:
        return self.model_settings | model_settings
    else:
        return self.model_settings or model_settings
```

Reply:
Good call, thanks!